actionlib/status.go

```diff
@@ -0,0 +1,224 @@
+package actionlib
+
+import (
+	"actionlib_msgs"
+	"fmt"
+	"hash/fnv"
+
+	"github.com/fetchrobotics/rosgo/ros"
+)
+
+type status struct {
+	as          *defaultActionServer
+	goal        ActionGoal
+	goalStatus  *actionlib_msgs.GoalStatus
+	destroyTime ros.Time
+}
+
+func newStatusWithActionGoal(as *defaultActionServer, goal ActionGoal) *status {
+	st := new(status)
+	st.as = as
+	st.goal = goal
+	timeNow := ros.Now()
+	st.destroyTime = timeNow.Add(as.statusListTimeout)
+	st.goalStatus = &actionlib_msgs.GoalStatus{}
+	st.goalStatus.GoalId = goal.GetGoalId()
+	st.goalStatus.Status = actionlib_msgs.PENDING
+
+	if st.goalStatus.GoalId.Id == "" {
+		st.goalStatus.GoalId.Id = fmt.Sprintf("nodeName-%d-%d-%d",
+			len(as.statusList), timeNow.Sec, timeNow.NSec)
+		st.goalStatus.GoalId.Stamp = timeNow
+	}
+
+	if st.goalStatus.GoalId.Stamp.IsZero() {
+		st.goalStatus.GoalId.Stamp = timeNow
+	}
+
+	return st
+}
+
+func newStatusWithGoalStatus(as *defaultActionServer, goalStatus actionlib_msgs.GoalStatus) *status {
+	st := new(status)
+	st.as = as
+	st.goalStatus = &goalStatus
+	timeNow := ros.Now()
+	st.destroyTime = timeNow.Add(as.statusListTimeout)
+	return st
+}
+
+func (st *status) setAccepted(text string) error {
+	if st.goal == nil {
+		return fmt.Errorf("attempt to set status on an uninitialized status handler")
+	}
+
+	status := st.goalStatus.Status
```

Comment:
It seems that we don't have this lock https://github.com/ros/actionlib/blob/indigo-devel/src/actionlib/server_goal_handle.py#L70? Would it cause any issue?

Reply:
Yes, it would, since it's being shared between action server and simple action server.